models.py

```diff
@@ -456,6 +461,47 @@ def sample_rate(self):
 	def read_audio(self):
 		return 'SoxAug' not in self.waveform_transform.__class__.__name__
 
+class Wav2VecFrontend(nn.Module):
+	def __init__(self,
+				 out_channels,
+				 *args,
+				 preemphasis = 0.0,
+				 use_context_features = True,
+				 fairseq_args = None,
+				 **kwargs):
+		from fairseq.models.wav2vec import Wav2VecModel
+
+		if fairseq_args.aggregator == 'cnn':
+			agg_layers = eval(fairseq_args.conv_aggregator_layers)
+			agg_dim = agg_layers[-1][0]
+			assert out_channels == agg_dim, f'Out channels {out_channels} is not equal to frontend output dim {agg_dim}, use --num-input-features {agg_dim}'
+		elif fairseq_args.aggregator == 'gru':
+			assert out_channels == fairseq_args.gru_dim, f'Out channels {out_channels} is not equal to frontend output dim {fairseq_args.gru_dim}, use --num-input-features {fairseq_args.gru_dim}'
+		else:
+			raise RuntimeError(f'Wrong wav2vec aggregator {fairseq_args.aggregator}. Use cnn or gru instead.')
+
+		super().__init__()
+		self.fairseq_args = fairseq_args
+		self.preemphasis = preemphasis
+		self.use_context_features = use_context_features
+		self.model = Wav2VecModel.build_model(fairseq_args, None).eval()
+
+	def forward(self, signal : shaping.BT, mask : shaping.BT = None) -> shaping.BCT:
+		signal = signal if signal.is_floating_point() else signal.to(torch.float32)
```

Comment:
is this needed? i don't remember myself

Reply:
Made by analogy with LogFilterBankFrontend. I dont know.

Reply:
This was needed there:
```python
signal = signal if signal.is_floating_point() else signal.to(torch.float32)
signal = normalize_signal(signal) if self.normalize_signal else signal
```
because signal was potentially unnormalized (and in int16 format). If you do not support int16 signals and assume float32 inputs, it's better to put at assert and remove the cast (the cast could hide a data loading mistake)